Spring Boot practice. Cart 1.0

aksoli666 opened the pull request:
added ShoppingCart and CartItem entities
added repository and service layers, also controller layer

aksoli666 committed:
Spring Boot practice. Cart 1.0
added ShoppingCart and CartItem entities
added repository and service layers, also controller layer

Elena-Bruyako commented on src/main/java/shop/controller/BookController.java:
```diff
@@ -36,7 +36,7 @@ public class BookController {
             description = "Create a new book"
     )
-    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    //@PreAuthorize("hasRole('ROLE_ADMIN')")
```
remove commented code everywhere

Elena-Bruyako commented on src/main/java/shop/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,68 @@
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping("/{id}")
+    public ShoppingCartResponseDto addBookToShoppingCart(
+            @PathVariable Long id, @RequestBody AddBookToCartRequestDto requestDto) {
```
```suggestion
            Authentication authentication, @RequestBody @Valid AddBookToCartRequestDto requestDto) {
```

Elena-Bruyako commented on src/main/java/shop/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,68 @@
+    )
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping("/items/{cartItemId}")
+    public CartItemResponseDto updateQuantityBook(
```
```suggestion
    public ShoppingCartResponseDto updateQuantityBook(
```

Elena-Bruyako commented on src/main/java/shop/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,68 @@
+            description = "Remove book from shopping cart"
+    )
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping("/items/{cartItemId}/{bookId}")
```
```suggestion
    @DeleteMapping("/items/{cartItemId}")
```

Elena-Bruyako commented on src/main/java/shop/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,68 @@
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping("/items/{cartItemId}/{bookId}")
+    public void removeBookFromShoppingCart(
+            @PathVariable Long cartItemId, @PathVariable Long bookId) {
```
```suggestion
            Authentication authentication,
                                   @PathVariable @Positive Long cartItemId) {
```

Elena-Bruyako commented on src/main/java/shop/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,68 @@
+            description = "Update quantity book in shopping cart"
+    )
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping("/items/{cartItemId}")
```
```suggestion
    @PutMapping("/items/{cartItemId}")
```

Elena-Bruyako commented on src/main/java/shop/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,68 @@
+    @PostMapping("/items/{cartItemId}")
+    public CartItemResponseDto updateQuantityBook(
+            @PathVariable Long cartItemId,
+            @RequestBody UpdateQuantityBookRequestDto requestDto) {
```
```suggestion
            Authentication authentication,
            @RequestBody @Valid UpdateQuantityBookRequestDto requestDto) {
```

Elena-Bruyako commented on src/main/java/shop/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,68 @@
+            description = "Add book to shopping cart"
+    )
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping("/{id}")
```
```suggestion
    @PostMapping
```

Elena-Bruyako commented on src/main/java/shop/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,68 @@
+            description = "Retrieve a cart details by id"
+    )
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/{id}")
```
```suggestion
    @GetMapping
```

Elena-Bruyako commented on src/main/java/shop/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,68 @@
+    )
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/{id}")
+    public ShoppingCartResponseDto getShoppingCartById(@PathVariable Long id) {
```
```suggestion
    public ShoppingCartResponseDto getShoppingCartById(Authentication authentication) {
```

Elena-Bruyako commented on src/main/java/shop/dto/request/shopping/cart/AddBookToCartRequestDto.java:
```diff
@@ -0,0 +1,13 @@
+@Getter
+@Setter
+public class AddBookToCartRequestDto {
+    @NotNull
+    private Long bookId;
+    private int quantity;
```
```suggestion
    @NotNull
    @Positive
    private Long bookId;
    @Positive
    private int quantity;
```

Elena-Bruyako commented on src/main/java/shop/dto/request/shopping/cart/UpdateQuantityBookRequestDto.java:
```diff
@@ -0,0 +1,10 @@
+@Getter
+@Setter
+public class UpdateQuantityBookRequestDto {
+    private int quantity;
```
```suggestion
    @Positive
    private int quantity;
```

Elena-Bruyako commented on src/main/java/shop/entity/CartItem.java:
```diff
@@ -0,0 +1,36 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

Elena-Bruyako commented on src/main/java/shop/entity/CartItem.java:
```diff
@@ -0,0 +1,36 @@
+    @ManyToOne
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

Elena-Bruyako commented on src/main/java/shop/entity/ShoppingCart.java:
```diff
@@ -0,0 +1,42 @@
+    @OneToOne
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @OneToMany
+    @JoinTable(
+            name = "carts_items",
+            joinColumns = @JoinColumn(name = "shopping_cart_id"),
+            inverseJoinColumns = @JoinColumn(name = "cart_item_id")
+    )
```
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

Elena-Bruyako commented on src/main/java/shop/entity/ShoppingCart.java:
```diff
@@ -0,0 +1,42 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
```
You can use `@MapsId` here

Elena-Bruyako commented on src/main/java/shop/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,68 @@
+@RequiredArgsConstructor
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
```
Create separate method to get `Long userId` from `Authentication authentication` and provide it to each Service method
`User user = (User) authentication.getPrincipal();`

Elena-Bruyako commented on src/main/java/shop/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+    @Transactional
+    @Override
+    public ShoppingCartResponseDto getShoppingCartById(Long id) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartById(id)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can`t get shopping cart by id: " + id));
+        Set<CartItemResponseDto> cartItemResponseDtos = cartItemMapper
+                .toCartItemResponseDtos(shoppingCart.getCartItems());
+        ShoppingCartResponseDto shoppingCartResponseDto = shoppingCartMapper
+                .toShoppingCartResponseDto(shoppingCart);
+        shoppingCartResponseDto.setCartItems(cartItemResponseDtos);
+        return shoppingCartResponseDto;
```
Example of implementation

```suggestion
        return shoppingCartMapper.toDto(shoppingCartRepository.findByUserId(userId));
```

Elena-Bruyako commented on src/main/java/shop/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+
+    @Transactional
```
remove

Elena-Bruyako commented on src/main/java/shop/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+    @Override
+    public CartItemResponseDto updateQuantityBook(Long cartItemId,
+                                                  UpdateQuantityBookRequestDto requestDto) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
```
Example of implementation
find shop cart
find cart Item `findByIdAndShoppingCartId`
change quantity
save cartItem

Elena-Bruyako commented on src/main/java/shop/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+        return shoppingCartResponseDto;
+    }
+
+    @Override
```
```suggestion
    @Override
    @Transactional
```

Elena-Bruyako commented on src/main/java/shop/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+    @Transactional
+    @Override
+    public void removeBookFromShoppingCart(Long cartItemId, Long bookId) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
```
You can implement this method in this way: 

find shop cart
find cartItem `findByIdAndShoppingCartId`
create method `removeItemFromCart(cartItem);` in ShoppingCart entity
```
public void removeItemFromCart(CartItem cartItem) {
        cartItems.remove(cartItem);
    }
```
call it

Elena-Bruyako commented on src/main/java/shop/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+        return cartItemMapper.toCartItemResponseDto(cartItemRepository.save(updated));
+    }
+
+    @Transactional
```
```suggestion
```

Elena-Bruyako commented on src/main/java/shop/service/impl/UserServiceImpl.java:
```diff
@@ -34,6 +37,9 @@ public UserResponseDto register(UserRegistrationRequestDto registerUserRequestDt
         user.setRoles(Set.of(roleRepository.findByRole(Role.RoleName.ROLE_USER)
                 .orElseThrow(() -> new EntityNotFoundException("Can`t find role"))));
         userRepository.save(user);
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
```
create and implement `createShoppingCart` method in ShopCartService

```suggestion
        shoppingCartService.createShoppingCart(user);
```

Elena-Bruyako reviewed (changes requested):
I didn't comment all places because there are a lot of changes should be done. So let's fix comments and then review again.

aksoli666 committed:
Spring Boot practice. Cart 2.0
tried to correct all the mistakes;
*in general changes related to shoppingCartService

liliia-ponomarenko commented on src/main/java/shop/controller/BookController.java:
```diff
@@ -36,7 +37,7 @@ public class BookController {
             description = "Create a new book"
     )
-    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    //@PreAuthorize("hasRole('ROLE_ADMIN')")
```
not fixed

liliia-ponomarenko commented on src/main/java/shop/controller/BookController.java:
```diff
@@ -67,9 +68,9 @@ public Page<BookDto> getAllBooks(Pageable pageable) {
             description = "Update a specific book"
     )
-    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    //@PreAuthorize("hasRole('ROLE_ADMIN')")
```
the same  (fix it everywhere)

liliia-ponomarenko commented on src/main/java/shop/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,72 @@
+            description = "Retrieve a cart details by id"
+    )
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping()
```
```suggestion
    @GetMapping
```

liliia-ponomarenko commented on src/main/java/shop/service/impl/UserServiceImpl.java:
```diff
@@ -34,6 +36,7 @@ public UserResponseDto register(UserRegistrationRequestDto registerUserRequestDt
         user.setRoles(Set.of(roleRepository.findByRole(Role.RoleName.ROLE_USER)
                 .orElseThrow(() -> new EntityNotFoundException("Can`t find role"))));
         userRepository.save(user);
+        shoppingCartService.createShoppingCart(user);
```
Let’s use Transactional for this method ;)

liliia-ponomarenko commented on src/main/java/shop/mapper/ShoppingCartMapper.java:
```diff
@@ -0,0 +1,19 @@
+import shop.dto.responce.shopping.cart.ShoppingCartResponseDto;
+import shop.entity.ShoppingCart;
+
+@Mapper(config = MapperConfig.class)
```
```suggestion
@Mapper(config = MapperConfig.class, uses = CartItemMapper.class)
```

liliia-ponomarenko commented on src/main/java/shop/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Shopping cart not found by user id: " + userId));
+        Set<CartItemResponseDto> cartItemResponseDtos = cartItemMapper
+                .toCartItemResponseDtos(shoppingCart.getCartItems());
+        ShoppingCartResponseDto shoppingCartResponseDto = shoppingCartMapper
+                .toShoppingCartResponseDto(shoppingCart);
+        shoppingCartResponseDto.setCartItems(cartItemResponseDtos);
```
you can simplify this mapping (follow my comment in the mapper). Let’s discover how it works ;)

liliia-ponomarenko commented on src/main/java/shop/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can`t get shopping cart for user with id : " + userId));
+        Set<CartItem> cartItems = shoppingCart.getCartItems();
+        CartItem existingCartItem = cartItems.stream()
+                .filter(cartItem -> cartItem.getBook().getId().equals(requestDto.getBookId()))
+                .findFirst()
+                .orElse(null);
+        if (existingCartItem != null) {
+            existingCartItem.setQuantity(requestDto.getQuantity() + existingCartItem.getQuantity());
+        } else {
+            CartItem cartItem = new CartItem();
+            cartItem.setShoppingCart(shoppingCart);
+            cartItem.setQuantity(requestDto.getQuantity());
+            cartItem.setBook(bookRepository.findById(requestDto.getBookId())
+                    .orElseThrow(() -> new EntityNotFoundException("Can`t get book by id: "
+                            + requestDto.getBookId())));
+            cartItems.add(cartItem);
+            cartItemRepository.save(cartItem);
+        }
+        shoppingCart.setCartItems(cartItems);
```
Since you have separate methods for adding a book to the cart and updating the number of books, you don't need to write such complex logic. Unless you can check whether there is no such book in the cart and throw out an error that such a book has already been added to the cart (don't forget to return the corresponding status code to the user)

liliia-ponomarenko commented on src/main/java/shop/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+        }
+        shoppingCart.setCartItems(cartItems);
+        shoppingCart = shoppingCartRepository.save(shoppingCart);
+        ShoppingCartResponseDto shoppingCartResponseDto = shoppingCartMapper
+                .toShoppingCartResponseDto(shoppingCart);
+        shoppingCartResponseDto.setCartItems(cartItemMapper.toCartItemResponseDtos(cartItems));
```
Let’s simplify this mapping ;)

liliia-ponomarenko commented on src/main/java/shop/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+        cartItem.setQuantity(cartItem.getQuantity() + requestDto.getQuantity());
+        cartItemRepository.save(cartItem);
+        ShoppingCart shoppingCart = cartItem.getShoppingCart();
+        ShoppingCartResponseDto shoppingCartResponseDto = shoppingCartMapper
+                .toShoppingCartResponseDto(shoppingCart);
+        shoppingCartResponseDto.setCartItems(
+                cartItemMapper.toCartItemResponseDtos(shoppingCart.getCartItems()));
```
the same

liliia-ponomarenko commented on src/main/java/shop/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't get cart item by id: " + cartItemId));
+        ShoppingCart shoppingCart = cartItem.getShoppingCart();
+        if (cartItem.getQuantity() <= 1) {
+            shoppingCart.removeItemFromCart(cartItem);
+            cartItemRepository.delete(cartItem);
+            if (shoppingCart.getCartItems().isEmpty()) {
+                shoppingCart.setCartItems(new HashSet<>());
+            }
+        } else {
+            cartItem.setQuantity(cartItem.getQuantity() - 1);
+            cartItemRepository.save(cartItem);
+        }
```
Let's remove this logic, this method involves completely removing the product from the cart. You have an update method to change the quantity. For a better understanding, open the website of any store. What requests are sent to the API under the hood of each button? This method will be called when the user clicks the remove button on the corresponding item in the cart.

liliia-ponomarenko commented on src/main/resources/db/changelog/db.changelog-master.yaml:
```diff
@@ -13,3 +13,10 @@ databaseChangeLog:
+      file: db/changelog/changes/09-create-cart-items-table.yaml
+  - include:
+      file: db/changelog/changes/10-create-carts-items-table.yaml
+
```
```suggestion
```
only one empty line ;)

liliia-ponomarenko reviewed (changes requested):
Good job! Let’s improve your solution ;)

aksoli666 committed:
Spring Boot practice. Cart 3.0
removed the annotation @PreAuthorize comment
used transactional
used CartItemMapper in shopping cart mapper
improved methods in shoppingCartService
deleted empty line in change log master
